File: Wells/Source/Filters/DateRangeFilter.vb
Imports Wells.CorePersistence.Repositories

Public Class DateRangeFilter(Of T)
    Inherits BaseFilter(Of T)

    Property StartDate As Date

    Property EndDate As Date

    Public Overrides ReadOnly Property DisplayValue As String
        Get
            Return $"Entre {StartDate.ToString("dd/MM/YY")} y {EndDate.ToString("dd/MM/YY")}"
        End Get
    End Property

    Public Overrides ReadOnly Property IsDateRangeFilter As Boolean
        Get
            Return True
        End Get
    End Property

    Sub New(propertyName As String, displayName As String, repo As IBussinessObjectRepository, startDate As Date)
        MyBase.New(propertyName, displayName, repo)
        Me.StartDate = startDate
        EndDate = Date.Today
    End Sub

    Sub New(propertyName As String, displayName As String, repo As IBussinessObjectRepository, startDate As Date, endDate As Date)
        Me.New(propertyName, displayName, repo, startDate)
        Me.EndDate = endDate
    End Sub

    Public Overrides Function Apply(originalList As IQueryable(Of T)) As IQueryable(Of T)
        Return From o In originalList
               Where CompareValue(CType(CallByName(o, PropertyName, CallType.Get), Date))
               Select o
    End Function

    Private Function CompareValue(objectValue As Date) As Boolean
        Return objectValue >= StartDate AndAlso objectValue <= EndDate
    End Function

    Overrides ReadOnly Property Description As String
        Get
            Return $"{DisplayPropertyName} entre {StartDate.ToString("dd/MM/YY")} y {EndDate.ToString("dd/MM/YY")}"
        End Get
    End Property

End Class

File: Wells/src/Filters/DateRangeFilter.cs
using Microsoft.VisualBasic;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Xml;
using Wells.Base;
using Wells.View.ViewModels;

namespace Wells.View.Filters
{
    public class DateRangeFilter<T> : BaseFilter<T>
    {
        public DateTime StartDate { get; set; }
        public DateTime EndDate { get; set; }

        public override string DisplayValue => $"Entre {StartDate.ToString("dd/MM/yy")} y {EndDate.ToString("dd/MM/yy")}";

        public override bool IsDateRangeFilter => true;

        public DateRangeFilter(string propertyName, string displayName, DateTime startDate) :
            base(propertyName, displayName)
        {
            StartDate = startDate;
            EndDate = DateTime.Today;
        }

        public DateRangeFilter() { }

        public DateRangeFilter(string propertyName, string displayName, DateTime startDate, DateTime endDate) :
           base(propertyName, displayName)
        {
            StartDate = startDate;
            EndDate = endDate;
        }

        public override IEnumerable<T> Apply(IEnumerable<T> queryable)
        {
            return from o in queryable
                   where CompareValue((DateTime)Interaction.CallByName(o, PropertyName, CallType.Get))
                   select o;
        }

        bool CompareValue(DateTime objectValue)
        {
            return objectValue >= StartDate && objectValue <= EndDate;
        }

        public override void SetUpdatedValues(FilterViewModel filterViewModel)
        {
            StartDate = filterViewModel.StartDate;
            EndDate = filterViewModel.EndDate;
        }

        public override string Description => $"{DisplayPropertyName} entre {StartDate.ToString("dd/MM/yy")} y {EndDate.ToString("dd/MM/yy")}";

        public override void ReadXml(XmlReader reader)
        {
            base.ReadXml(reader);
            StartDate = reader.ReadContentAsDateTime();
            EndDate = reader.ReadContentAsDateTime();
            reader.ReadEndElement();
        }

        public override void WriteXml(XmlWriter writer)
        {
            base.WriteXml(writer);
            writer.Write(nameof(StartDate), StartDate);
            writer.Write(nameof(EndDate), EndDate);
        }
    }
}
